include/rx/ranges.hpp

```diff
@@ -1354,62 +1620,96 @@ struct foldl {
 
     template <class InputRange>
     constexpr T operator()(InputRange&& input) {
+        // Copying for reentrancy.
         auto copy = as_input_range(std::forward<InputRange>(input));
         RX_TYPE_ASSERT(is_finite<decltype(copy)>);
         T accum = init;
-        while (!copy.at_end()) {
+        if (RX_UNLIKELY(copy.at_end()))
+            return accum;
+        do {
             accum = func(accum, copy.get());
             copy.next();
-        }
+        } while (RX_LIKELY(!copy.at_end()));
         return accum;
     }
 };
 template <class T, class F>
-foldl(T&&, F &&)->foldl<RX_REMOVE_CVREF_T<T>, RX_REMOVE_CVREF_T<F>>;
+foldl(T&&, F &&)->foldl<remove_cvref_t<T>, remove_cvref_t<F>>;
+
+/*!
+    @brief Sum the input range.
 
+    The output type of the input range is required to implement `operator+`.
+*/
 struct sum {
     constexpr sum() noexcept {}
```

Comment:
It should not be defined or defaulted to keep the constructor trivial.

Reply:
:+1: